clippy_utils/src/qualify_min_const_fn.rs

```diff
@@ -80,6 +81,27 @@ fn check_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, span: Span) -> McfResult {
             GenericArgKind::Lifetime(_) | GenericArgKind::Const(_) => continue,
         };
 
+        // Only do this check if we're in `TerminatorKind::Drop`, otherwise rustc will sometimes overflow
+        // its stack. This check is unnecessary outside of a `Drop` anyway so it's faster regardless
+        if in_drop && let ty::Adt(def, subst) = ty.kind() {
+            if def.has_non_const_dtor(tcx) && in_drop {
```

Comment:
Is there really a need to walk all the field types after this check?

Reply:
AFAIK yes. from my testing it wouldn't catch it's not `~const Destruct` without it, but I can test again.

Reply:
That's true. It doesn't take `ParamEnv`.

Reply:
It would be better to use the trait solver here rather than walking the type. I would make a new util function for it. You can use [rustc's impl](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_const_eval/transform/check_consts/qualifs.rs.html#150) as a reference.

Reply:
> It would be better to use the trait solver here rather than walking the type.

I'm trying to do this but it seems that it returns true anyway, which is... confusing, since it checks `Vec` and all but still thinks it has `~const Destruct` implemented...

Reply:
It seems `Vec<u16>` returns `Err(Unimplemented)` so I'll likely still need to walk the type... Am I missing something?

Reply:
Well, I just committed the current implementation and it's slow, really slow. And it still needs to walk all fields so I don't think it should do this

Reply:
You didn't use any of the reference function starting at line 173 and that looks like it's filtering out unusable candidate selections.

Reply:
I actually did before; I removed it because it didn't change the result at all

Reply:
That is the case. It wouldn't be too bad to do so now as const impls are still unstable.

Now that I'm looking at it, the referenced issue doesn't have a `Drop` terminator in the MIR at all. Which makes sense as the only contained field was moved.

Reply:
I'm still trying to fix this but it seems like this won't work (for both `Destruct` and also `Drop`), as it seems nothing checks whether its fields implement `~const Destruct` or not; just the containing type, therefore no matter what it needs to walk it.

If that's the case, I think the original implementation should be used since it's a lot faster...

Reply:
It's worth noting that using the trait solver works if the containing type doesn't implement `~const Destruct`, just not if only the fields don't

Reply:
Wow nevermind, I got it working... I have no clue why `ocx.select_all_or_error()` didn't work before but it's working so I'll take it? That's so bizarre